go/pkg/reader/reader.go

```diff
@@ -187,55 +193,34 @@ func NewCompressedSeeker(fs ReadSeeker) (ReadSeeker, error) {
 		fs:    fs,
 		encdW: encdW,
 		buf:   sb,
-		err:   nil,
 	}, nil
 }
 
 func (cfs *compressedSeeker) Read(p []byte) (int, error) {
-	if cfs.err != nil {
-		return 0, cfs.err
-	}
-
-	var n int
-	var errR, errW error
-	for cfs.encdW != nil && cfs.buf.Len() < len(p) && errR == nil && errW == nil {
-		// Read is allowed to use the entity of p as a scratchpad.
-		n, errR = cfs.fs.Read(p)
-		// errW must be non-nil if written bytes != from n.
-		_, errW = cfs.encdW.Encoder.Write(p[:n])
-	}
-
-	// We have to treat EOF differently. It's the only "everything is going according
-	// to plan" error. EOF implies our uncompressed input data has finished reading,
-	// but it doesn't mean that we've finished preparing/returning our compressed data.
-	// We only want to return EOF to the user when the _compressed_ data is finished reading.
-	if errR != nil && errR != io.EOF {
-		cfs.err = errR
-	} else if errW != nil {
-		cfs.err = errW
-	}
-
-	// When the buffer ends, or in case of _any_ error, we compress all the bytes we
-	// had available. The encoder requires a Close call to finish writing compressed
-	// data smaller than zstd's window size.
-	var errC error
-	if cfs.err != nil || errR == io.EOF {
-		errC = cfs.encdW.Encoder.Close()
+	if !cfs.IsInitialized() {
+		return 0, errNotInitialized
+	}
+
+	var err error
+	if cfs.encdW != nil && cfs.buf.Len() < len(p) {
```

Comment:
Do you know why the check `cfs.buf.Len() < len(p)` matters?

Reply:
I believe it's an optimization – if there's already enough cached compressed data available to fill `p`, then we can skip compressing more data and just use the already-compressed data.

I think that only happens when the compressed data is *longer* than the noncompressed data (since that's the only time that a previous `Read()` call might leave some data in the buffer). A comment in the test file says that can happen if the size of the data is very small due to the addition of headers.

Reply:
Ah, I see. Seems like this implementation is not as efficient as it could be due to this staging. Oh well.